scripts/platoon-launcher

```diff
@@ -38,28 +38,37 @@ if __name__ == '__main__':
     os.makedirs(logs_folder)
 
     print "### Launching experiment: {}".format(args.experiment_name)
-    process_list = {}
+    process_map = {}
 
-    process_list['Controller'] = launch_process(logs_folder, args.experiment_name, args.controller_args, "cpu", "controller")
+    p = launch_process(logs_folder, args.experiment_name, args.controller_args, "cpu", "controller")
+    process_map[p.pid] = ('Controller', p)
 
     for device in args.gpu_list:
         worker_process = launch_process(logs_folder, args.experiment_name, args.workers_args, device)
-        process_list["Worker{}".format(device)] = worker_process
+        process_map[worker_process.pid] = ("Worker{}".format(device),
+                                           worker_process)
 
     print "\n### Logs folder ###\n{}".format(logs_folder)
 
     print "\n### Waiting on experiment to finish ..."
 
     # Silly error handling but that will do for now.
-    for name in process_list:
-        returncode = process_list[name].wait()
-        print "{} terminated with return code: {}.".format(name, returncode)
-
+    while process_map:
+        pid, returncode = os.wait()
+        if pid not in process_map:
+            print "Recieved status for unknown process {}".format(pid)
+
+        name, p = process_map[pid]
+        del process_map[pid]
+        print "{} terminated with return code: {}.".format(name,
+                                                           returncode)
         if returncode != 0:
             print "\nWARNING! An error has occurred.\nCleaning up and closing, see logs folder."
-            for p in process_list:
-                try:
-                    process_list[p].kill()
-                except:
-                    pass
-            break
+            while process_map:
+                for name, p in list(process_map.keys()):
```

Comment:
I think he has list() there to force a copy, because iterating on a dict (or Python 3 KeysView) while modifying the keys in the dictionary will lead to errors about the dictionary changing size.